Fail navigations to about: scheme if the URL is not "about:blank" or "about:srcdoc"

cdumez opened the pull request:
#### 3c3163e71f647db507949ecebad35d0f2ffb33f3
<pre>
Fail navigations to about: scheme if the URL is not &quot;about:blank&quot; or &quot;about:srcdoc&quot;
<a href="https://bugs.webkit.org/show_bug.cgi?id=259946">https://bugs.webkit.org/show_bug.cgi?id=259946</a>
rdar://57966056

Reviewed by Brent Fulgham.

Fail navigations to about: scheme if the URL is not &quot;about:blank&quot; or &quot;about:srcdoc&quot;.
This matches what Blink and Gecko are doing.

* Source/WebCore/loader/DocumentLoader.cpp:
(WebCore::DocumentLoader::startLoadingMainResource):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/Navigation.mm:
(TEST):

Canonical link: <a href="https://commits.webkit.org/267754@main">https://commits.webkit.org/267754@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/3041aee0ee28322ebfe8f81e2dec9aecbd01df6b

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/15680 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/15988 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/16360 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/17438 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/14714 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/15862 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/18960 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/16090 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/17245 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/15866 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/16343 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/13353 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/18184 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/13602 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/14155 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/21072 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/14603 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/14321 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/17580 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/14912 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/12646 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/14165 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/14001 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/4193 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/18530 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/14731 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

brentfulgham commented on LayoutTests/http/tests/security/about-url-host.html:
```diff
@@ -5,19 +5,22 @@
 
-    if (window.testRunner)
-        testRunner.notifyDone();
+onload = () => {
```
Nice clean-up!

brentfulgham commented on Source/WebCore/loader/DocumentLoader.cpp:
```diff
@@ -2105,6 +2105,11 @@ void DocumentLoader::startLoadingMainResource()
 
     Ref<DocumentLoader> protectedThis(*this);
 
+    if (m_request.url().protocolIsAbout() && !(m_request.url().isAboutBlank() || m_request.url().isAboutSrcDoc())) {
```
I wonder if there is loading spec language that covers this stuff. Should we create a WPT test for this at some point?

karlcow replied:
special schemes section does not mention `about:` in https://url.spec.whatwg.org/#url-miscellaneous
@annevk would know.

annevk replied:
https://fetch.spec.whatwg.org/#scheme-fetch defines that for fetching only "about:blank" should give you something. And navigation special cases `about:srcdoc` as far as I can tell (look at the caller of https://html.spec.whatwg.org/#create-navigation-params-from-a-srcdoc-resource). So this behavior is covered by the standard and correct.

brentfulgham reviewed (approved):
Looks great -- might be worth considering if we should create a new WPT test for this (if one doesn't exist). That's not relevant to this PR, just something to consider for the future.

karlcow commented on Source/WebCore/loader/DocumentLoader.cpp:
```diff
@@ -2105,6 +2105,11 @@ void DocumentLoader::startLoadingMainResource()
+    if (m_request.url().protocolIsAbout() && !(m_request.url().isAboutBlank() || m_request.url().isAboutSrcDoc())) {
+        cancelMainResourceLoad(frameLoader()->client().cannotShowURLError(m_request));
+        return;
+    }
```
@cdumez @brentfulgham will it have any consequences on 

https://github.com/WebKit/WebKit/blob/0b71df88d47241c1024084a8f4edb0410d847520/Source/WebCore/page/Quirks.cpp#L822-L845

The origin bug was
https://bugs.webkit.org/show_bug.cgi?id=199933

which was about dealing with `about:invalid#zClosureZ`.
I don't need if the Quirk is needed. I haven't checked yet.

But would this make the Quirk useless?

cdumez replied:
My patch doesn't have consequences on this quirk because this quirk gets applied *before* my new check.

So no, it wouldn't make this quirk useless. If anything, it would make the quirk more important because `about:invalid#zClosureZ` (which would get blocked by my patch) would get converted to `about:blank` (which is allowed to load).

karlcow replied:
thanks for the clarification.

karlcow reviewed (commented):
needs to test docs.google.com with and without Site Specific Hacks.
Or maybe it will have no impact.

cdumez committed:
Fail navigations to about: scheme if the URL is not "about:blank" or "about:srcdoc"
https://bugs.webkit.org/show_bug.cgi?id=259946
rdar://57966056

Reviewed by Brent Fulgham.

Fail navigations to about: scheme if the URL is not "about:blank" or "about:srcdoc".
This matches what Blink and Gecko are doing.

* Source/WebCore/loader/DocumentLoader.cpp:
(WebCore::DocumentLoader::startLoadingMainResource):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/Navigation.mm:
(TEST):

Canonical link: https://commits.webkit.org/267754@main
